File: sdks/official/fraiseql-php/src/ConfigLoader.php
<?php

declare(strict_types=1);

namespace FraiseQL;

/**
 * Utility class that loads `fraiseql.toml` and parses the [inject_defaults]
 * section, then configures the SchemaRegistry with the parsed values.
 *
 * Supports three TOML sections:
 * - [inject_defaults]           -> base defaults for all operations
 * - [inject_defaults.queries]   -> additional defaults for queries
 * - [inject_defaults.mutations] -> additional defaults for mutations
 *
 * Each section contains key = "value" pairs (e.g. tenant_id = "jwt:tenant_id").
 *
 * Usage:
 * ```php
 * ConfigLoader::load('fraiseql.toml');
 * // Registry now has inject defaults configured
 * ```
 */
final class ConfigLoader
{
    /**
     * Load a fraiseql.toml file and apply inject_defaults to the SchemaRegistry.
     *
     * @param string $path Path to the fraiseql.toml file
     * @return void
     *
     * @throws FraiseQLException If the file cannot be read
     */
    public static function load(string $path): void
    {
        $content = file_get_contents($path);
        if ($content === false) {
            throw new FraiseQLException("Failed to read config file: $path");
        }

        $sections = self::parseInjectDefaults($content);

        SchemaRegistry::getInstance()->setInjectDefaults(
            $sections['base'],
            $sections['queries'],
            $sections['mutations'],
        );
    }

    /**
     * Parse inject_defaults sections from TOML content using line-by-line parsing.
     *
     * @param string $content The TOML file content
     * @return array{base: array<string, string>, queries: array<string, string>, mutations: array<string, string>}
     */
    private static function parseInjectDefaults(string $content): array
    {
        $base = [];
        $queries = [];
        $mutations = [];

        $currentSection = null;
        $lines = explode("\n", $content);

        foreach ($lines as $line) {
            $trimmed = trim($line);

            // Skip empty lines and comments
            if ($trimmed === '' || str_starts_with($trimmed, '#')) {
                continue;
            }

            // Detect section headers
            if (str_starts_with($trimmed, '[')) {
                $currentSection = self::parseSectionHeader($trimmed);
                continue;
            }

            // Parse key = "value" pairs within inject_defaults sections
            if ($currentSection !== null && str_contains($trimmed, '=')) {
                $pair = self::parseKeyValue($trimmed);
                if ($pair !== null) {
                    match ($currentSection) {
                        'base' => $base[$pair[0]] = $pair[1],
                        'queries' => $queries[$pair[0]] = $pair[1],
                        'mutations' => $mutations[$pair[0]] = $pair[1],
                        default => null,
                    };
                }
            }
        }

        return ['base' => $base, 'queries' => $queries, 'mutations' => $mutations];
    }

    /**
     * Parse a TOML section header and return the inject_defaults sub-section name.
     *
     * @param string $header The section header line (e.g. "[inject_defaults.queries]")
     * @return string|null The sub-section name ('base', 'queries', 'mutations') or null
     */
    private static function parseSectionHeader(string $header): ?string
    {
        // Remove brackets and whitespace
        $section = trim($header, "[] \t");

        if ($section === 'inject_defaults') {
            return 'base';
        }

        if ($section === 'inject_defaults.queries') {
            return 'queries';
        }

        if ($section === 'inject_defaults.mutations') {
            return 'mutations';
        }

        return null;
    }

    /**
     * Parse a TOML key = "value" line.
     *
     * @param string $line The line to parse
     * @return array{0: string, 1: string}|null The [key, value] pair or null
     */
    private static function parseKeyValue(string $line): ?array
    {
        $eqPos = strpos($line, '=');
        if ($eqPos === false) {
            return null;
        }

        $key = trim(substr($line, 0, $eqPos));
        $value = trim(substr($line, $eqPos + 1));

        // Strip surrounding quotes (single or double)
        if (
            (str_starts_with($value, '"') && str_ends_with($value, '"'))
            || (str_starts_with($value, "'") && str_ends_with($value, "'"))
        ) {
            $value = substr($value, 1, -1);
        }

        return [$key, $value];
    }
}

File: sdks/official/fraiseql-java/src/main/java/com/fraiseql/core/ConfigLoader.java
package com.fraiseql.core;

import java.io.BufferedReader;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

/**
 * Loads FraiseQL configuration from a TOML file.
 * Extracts {@code [inject_defaults]} sections and applies them to the SchemaRegistry.
 *
 * <p>Supported TOML sections:
 * <pre>
 * [inject_defaults]
 * tenant_id = "jwt:tenant_id"
 *
 * [inject_defaults.queries]
 * viewer_id = "jwt:sub"
 *
 * [inject_defaults.mutations]
 * actor_id = "jwt:sub"
 * </pre>
 */
public class ConfigLoader {

    private ConfigLoader() {
        // Utility class
    }

    /**
     * Load configuration from a TOML file and apply inject defaults to the registry.
     *
     * @param tomlPath path to the fraiseql.toml file
     * @throws IOException if reading the file fails
     */
    public static void loadConfig(String tomlPath) throws IOException {
        Map<String, String> base = new HashMap<>();
        Map<String, String> queries = new HashMap<>();
        Map<String, String> mutations = new HashMap<>();

        String currentSection = null;

        try (BufferedReader reader = Files.newBufferedReader(Path.of(tomlPath))) {
            String line;
            while ((line = reader.readLine()) != null) {
                line = line.strip();

                // Skip comments and blank lines
                if (line.isEmpty() || line.startsWith("#")) {
                    continue;
                }

                // Section header
                if (line.startsWith("[")) {
                    currentSection = line.replace("[", "").replace("]", "").strip();
                    continue;
                }

                // Key = "value" pair
                if (currentSection != null && currentSection.startsWith("inject_defaults") && line.contains("=")) {
                    int eqIdx = line.indexOf('=');
                    String key = line.substring(0, eqIdx).strip();
                    String value = line.substring(eqIdx + 1).strip();

                    // Strip surrounding quotes
                    if (value.startsWith("\"") && value.endsWith("\"")) {
                        value = value.substring(1, value.length() - 1);
                    } else if (value.startsWith("'") && value.endsWith("'")) {
                        value = value.substring(1, value.length() - 1);
                    }

                    if ("inject_defaults".equals(currentSection)) {
                        base.put(key, value);
                    } else if ("inject_defaults.queries".equals(currentSection)) {
                        queries.put(key, value);
                    } else if ("inject_defaults.mutations".equals(currentSection)) {
                        mutations.put(key, value);
                    }
                }
            }
        }

        SchemaRegistry.getInstance().setInjectDefaults(base, queries, mutations);
    }
}
